### PR TITLE
Implements Dupn Swapn Txindex new opcodes

### PR DESCRIPTION
**Pull request by ajlopez**
Implement new VM opcodes:

DUPN
SWAPN
TXINDEX

according to RSK IPs 26 and 11

### CODE REVIEW

**Comment by rskops** [comment]
SonarQube analysis reported 398 issues
* ![BLOCKER][BLOCKER] 2 blocker
* ![CRITICAL][CRITICAL] 3 critical
* ![MAJOR][MAJOR] 159 major
* ![MINOR][MINOR] 158 minor
* ![INFO][INFO] 76 info

#### Top 10 issues

1. ![BLOCKER][BLOCKER] [TransactionExecutor.java#L357](https://github.com/rsksmart/rskj/blob/238c2681e7ac1947a0f0aaceb36243df0f057059/rskj-core/src/main/java/org/ethereum/core/TransactionExecutor.java#L357): Catch Exception instead of Throwable. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1181)
1. ![BLOCKER][BLOCKER] [ProgramInvokeImpl.java#L304](https://github.com/rsksmart/rskj/blob/238c2681e7ac1947a0f0aaceb36243df0f057059/rskj-core/src/main/java/org/ethereum/vm/program/invoke/ProgramInvokeImpl.java#L304): This class overrides "equals()" and should therefore also override "hashCode()". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1206)
1. ![CRITICAL][CRITICAL] [TransactionExecutor.java#L139](https://github.com/rsksmart/rskj/blob/238c2681e7ac1947a0f0aaceb36243df0f057059/rskj-core/src/main/java/org/ethereum/core/TransactionExecutor.java#L139): This line will not be executed conditionally; only the first line of this 3-line block will be. The rest will execute unconditionally. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2681)
1. ![CRITICAL][CRITICAL] [TransactionExecutor.java#L395](https://github.com/rsksmart/rskj/blob/238c2681e7ac1947a0f0aaceb36243df0f057059/rskj-core/src/main/java/org/ethereum/core/TransactionExecutor.java#L395): Cast one of the operands of this multiplication operation to a "long". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2184)
1. ![CRITICAL][CRITICAL] [VM.java#L315](https://github.com/rsksmart/rskj/blob/238c2681e7ac1947a0f0aaceb36243df0f057059/rskj-core/src/main/java/org/ethereum/vm/VM.java#L315): Cast one of the operands of this addition operation to a "long". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2184)
1. ![MAJOR][MAJOR] [BlockExecutor.java#L144](https://github.com/rsksmart/rskj/blob/238c2681e7ac1947a0f0aaceb36243df0f057059/rskj-core/src/main/java/co/rsk/core/bc/BlockExecutor.java#L144): Method co.rsk.core.bc.BlockExecutor.validate(Block, BlockResult) appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [BlockExecutor.java#L216](https://github.com/rsksmart/rskj/blob/238c2681e7ac1947a0f0aaceb36243df0f057059/rskj-core/src/main/java/co/rsk/core/bc/BlockExecutor.java#L216): Method co.rsk.core.bc.BlockExecutor.execute(Block, byte[], boolean, boolean) does not presize the allocation of a collection [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APSC_PRESIZE_COLLECTIONS)
1. ![MAJOR][MAJOR] [BlockExecutor.java#L248](https://github.com/rsksmart/rskj/blob/238c2681e7ac1947a0f0aaceb36243df0f057059/rskj-core/src/main/java/co/rsk/core/bc/BlockExecutor.java#L248): Method co.rsk.core.bc.BlockExecutor.execute(Block, byte[], boolean, boolean) does not presize the allocation of a collection [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APSC_PRESIZE_COLLECTIONS)
1. ![MAJOR][MAJOR] [PendingStateImpl.java#L188](https://github.com/rsksmart/rskj/blob/238c2681e7ac1947a0f0aaceb36243df0f057059/rskj-core/src/main/java/co/rsk/core/bc/PendingStateImpl.java#L188): Method co.rsk.core.bc.PendingStateImpl.addWireTransactions(List) does not presize the allocation of a collection [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APSC_PRESIZE_COLLECTIONS)
1. ![MAJOR][MAJOR] [TransactionExecutor.java#L60](https://github.com/rsksmart/rskj/blob/238c2681e7ac1947a0f0aaceb36243df0f057059/rskj-core/src/main/java/org/ethereum/core/TransactionExecutor.java#L60): Remove this unused "stateLogger" private field. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1068)

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by rskops** [comment]
SonarQube analysis reported 432 issues
* ![BLOCKER][BLOCKER] 2 blocker
* ![CRITICAL][CRITICAL] 3 critical
* ![MAJOR][MAJOR] 169 major
* ![MINOR][MINOR] 168 minor
* ![INFO][INFO] 90 info

#### Top 10 issues

1. ![BLOCKER][BLOCKER] [TransactionExecutor.java#L357](https://github.com/rsksmart/rskj/blob/9f02a0cc4570378674fe690694febc20c9bc3b4d/rskj-core/src/main/java/org/ethereum/core/TransactionExecutor.java#L357): Catch Exception instead of Throwable. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1181)
1. ![BLOCKER][BLOCKER] [ProgramInvokeImpl.java#L304](https://github.com/rsksmart/rskj/blob/9f02a0cc4570378674fe690694febc20c9bc3b4d/rskj-core/src/main/java/org/ethereum/vm/program/invoke/ProgramInvokeImpl.java#L304): This class overrides "equals()" and should therefore also override "hashCode()". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1206)
1. ![CRITICAL][CRITICAL] [TransactionExecutor.java#L139](https://github.com/rsksmart/rskj/blob/9f02a0cc4570378674fe690694febc20c9bc3b4d/rskj-core/src/main/java/org/ethereum/core/TransactionExecutor.java#L139): This line will not be executed conditionally; only the first line of this 3-line block will be. The rest will execute unconditionally. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2681)
1. ![CRITICAL][CRITICAL] [TransactionExecutor.java#L395](https://github.com/rsksmart/rskj/blob/9f02a0cc4570378674fe690694febc20c9bc3b4d/rskj-core/src/main/java/org/ethereum/core/TransactionExecutor.java#L395): Cast one of the operands of this multiplication operation to a "long". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2184)
1. ![CRITICAL][CRITICAL] [VM.java#L315](https://github.com/rsksmart/rskj/blob/9f02a0cc4570378674fe690694febc20c9bc3b4d/rskj-core/src/main/java/org/ethereum/vm/VM.java#L315): Cast one of the operands of this addition operation to a "long". [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS2184)
1. ![MAJOR][MAJOR] [BlockExecutor.java#L144](https://github.com/rsksmart/rskj/blob/9f02a0cc4570378674fe690694febc20c9bc3b4d/rskj-core/src/main/java/co/rsk/core/bc/BlockExecutor.java#L144): Method co.rsk.core.bc.BlockExecutor.validate(Block, BlockResult) appears to call the same method on the same object redundantly [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APRMC_POSSIBLY_REDUNDANT_METHOD_CALLS)
1. ![MAJOR][MAJOR] [BlockExecutor.java#L216](https://github.com/rsksmart/rskj/blob/9f02a0cc4570378674fe690694febc20c9bc3b4d/rskj-core/src/main/java/co/rsk/core/bc/BlockExecutor.java#L216): Method co.rsk.core.bc.BlockExecutor.execute(Block, byte[], boolean, boolean) does not presize the allocation of a collection [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APSC_PRESIZE_COLLECTIONS)
1. ![MAJOR][MAJOR] [BlockExecutor.java#L248](https://github.com/rsksmart/rskj/blob/9f02a0cc4570378674fe690694febc20c9bc3b4d/rskj-core/src/main/java/co/rsk/core/bc/BlockExecutor.java#L248): Method co.rsk.core.bc.BlockExecutor.execute(Block, byte[], boolean, boolean) does not presize the allocation of a collection [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APSC_PRESIZE_COLLECTIONS)
1. ![MAJOR][MAJOR] [PendingStateImpl.java#L188](https://github.com/rsksmart/rskj/blob/9f02a0cc4570378674fe690694febc20c9bc3b4d/rskj-core/src/main/java/co/rsk/core/bc/PendingStateImpl.java#L188): Method co.rsk.core.bc.PendingStateImpl.addWireTransactions(List) does not presize the allocation of a collection [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=fb-contrib%3APSC_PRESIZE_COLLECTIONS)
1. ![MAJOR][MAJOR] [TransactionExecutor.java#L60](https://github.com/rsksmart/rskj/blob/9f02a0cc4570378674fe690694febc20c9bc3b4d/rskj-core/src/main/java/org/ethereum/core/TransactionExecutor.java#L60): Remove this unused "stateLogger" private field. [rule](https://jenkins.rsklabs.io/sonar/coding_rules#rule_key=squid%3AS1068)

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: rskj-core/src/main/java/org/ethereum/core/TransactionExecutor.java [hunk 1]
```diff
@@ -89,17 +90,18 @@
...
-    public TransactionExecutor(Transaction tx, byte[] coinbase, Repository track, BlockStore blockStore, ReceiptStore receiptStore,
+    public TransactionExecutor(Transaction tx, int txindex, byte[] coinbase, Repository track, BlockStore blockStore, ReceiptStore receiptStore,
                                ProgramInvokeFactory programInvokeFactory, Block executionBlock) {
 
-        this(tx, coinbase, track, blockStore, receiptStore, programInvokeFactory, executionBlock, new EthereumListenerAdapter(), 0);
+        this(tx, txindex, coinbase, track, blockStore, receiptStore, programInvokeFactory, executionBlock, new EthereumListenerAdapter(), 0);
     }
 
-    public TransactionExecutor(Transaction tx, byte[] coinbase, Repository track, BlockStore blockStore, ReceiptStore receiptStore,
+    public TransactionExecutor(Transaction tx, int txindex, byte[] coinbase, Repository track, BlockStore blockStore, ReceiptStore receiptStore,
```
Comment (donequis) [inline rskj-core/src/main/java/org/ethereum/core/TransactionExecutor.java]: If txindex is always 0 you can omit this parameter

File: rskj-core/src/main/java/org/ethereum/vm/OpCode.java [hunk 2]
```diff
@@ -611,7 +626,7 @@
      */
     SUICIDE(0xff, 1, 0, ZERO_TIER);
 
-    public final byte opcode;
+    public final byte opCode;
```
Comment (donequis) [inline rskj-core/src/main/java/org/ethereum/vm/OpCode.java]: Useless rename. Also opcode is one word.

File: rskj-core/src/main/java/org/ethereum/vm/VM.java [hunk 3]
```diff
@@ -942,10 +955,40 @@ protected void doDUP() {
...
 
+    protected void doDUPN() {
+        spendOpCodeGas();
+        // EXECUTION PHASE
+        program.step();
+
+        int n = stack.pop().intValueCheck() + 1;
+
+        program.verifyStackSize(n);
+        program.verifyStackOverflow(n, n + 1);
```
Comment (donequis) [inline rskj-core/src/main/java/org/ethereum/vm/VM.java]: Stack is checked on steps()

Comment (donequis) [inline rskj-core/src/main/java/org/ethereum/vm/VM.java]: Stack is checked at steps(), and checking implementaitons it adds nothing doing this here. We can add 0, 1 for stack overflow.

File: rskj-core/src/main/java/org/ethereum/vm/VM.java [hunk 4]
```diff
@@ -942,10 +955,40 @@ protected void doDUP() {
...
+
+    protected void doSWAPN(){
+        spendOpCodeGas();
+        // EXECUTION PHASE
+        program.step();
+
+        int n = stack.pop().intValueCheck() + 2;
+
+        program.verifyStackSize(n);
+        program.verifyStackOverflow(n, n);
```
Comment (donequis) [inline rskj-core/src/main/java/org/ethereum/vm/VM.java]: Useless call

File: rskj-core/src/main/java/org/ethereum/vm/VM.java [hunk 5]
```diff
@@ -1619,6 +1663,8 @@ protected void executeOpcode() {
             case OpCodes.OP_SWAP_15:
             case OpCodes.OP_SWAP_16: doSWAP();
             break;
+            case OpCodes.OP_SWAPN: doSWAPN();
+                break;
```
Comment (donequis) [inline rskj-core/src/main/java/org/ethereum/vm/VM.java]: Indentation

File: rskj-core/src/main/java/org/ethereum/vm/VM.java [hunk 6]
```diff
@@ -1693,6 +1739,8 @@ protected void executeOpcode() {
             break;
             case OpCodes.OP_CODEREPLACE: doCODEREPLACE();
             break;
+            case OpCodes.OP_DUPN: doDUPN();
+                break;
```
Comment (donequis) [inline rskj-core/src/main/java/org/ethereum/vm/VM.java]: Indentation

**Pull request closed without merging**